rust/connlib/tunnel/src/client.rs

```diff
@@ -281,37 +281,61 @@ fn send_dns_answer(
     }
 }
 
+/// A SANS-IO implementation of a client's functionality.
+///
+/// Internally, this composes a [`snownet::ClientNode`] with firezone's policy engine around resources.
+/// Clients differ from gateways in that they also implement a DNS resolver for DNS resources.
+/// They also initiate connections to gateways based on packets sent to resources.
 pub struct ClientState {
+    /// The [`snownet::ClientNode`].
+    ///
+    /// Manages wireguard tunnels to gateways.
+    node: ClientNode<GatewayId, RelayId>,
+    /// All peers gateways we are connected to and the associated, connection-specific state.
+    peers: PeerStore<GatewayId, GatewayOnClient>,
+    /// Which resources are trying to connect to.
     awaiting_connection: HashMap<ResourceId, AwaitingConnectionDetails>,
+
+    /// Tracks, which gateway to use for a particular resource.
     resources_gateways: HashMap<ResourceId, GatewayId>,
+    /// The site a gateway belongs to.
+    gateways_site: HashMap<GatewayId, SiteId>,
+    /// The online/offline status of a site.
+    sites_status: HashMap<SiteId, Status>,
 
-    pub dns_resources_internal_ips: HashMap<DnsResource, HashSet<IpAddr>>,
+    /// All DNS resources we know about, indexed by their domain (could be wildcard domain like *.mycompany.com).
     dns_resources: HashMap<String, ResourceDescriptionDns>,
+    /// All CIDR resources we know about, indexed by the IP range they cover (like `1.1.0.0/8`).
     cidr_resources: IpNetworkTable<ResourceDescriptionCidr>,
-    pub resource_ids: HashMap<ResourceId, ResourceDescription>,
-    pub deferred_dns_queries: HashMap<(DnsResource, Rtype), IpPacket<'static>>,
-
-    peers: PeerStore<GatewayId, GatewayOnClient>,
-
-    node: ClientNode<GatewayId, RelayId>,
-
-    pub ip_provider: IpProvider,
-
-    dns_mapping: BiMap<IpAddr, DnsServer>,
+    /// All resources indexed by their ID.
+    resource_ids: HashMap<ResourceId, ResourceDescription>,
 
-    buffered_events: VecDeque<ClientEvent>,
-    interface_config: Option<InterfaceConfig>,
-    buffered_packets: VecDeque<IpPacket<'static>>,
+    /// The DNS resolvers configured on the system prior to connlib starting up.
```

Comment:
```suggestion
    /// The DNS resolvers configured on the system outside of connlib
```

I think it's only the headless Clients that don't support refreshing DNS servers yet, and that may be something admins want eventually so they can move towards zero downtime.